lib/spack/spack/environment.py

```diff
@@ -284,132 +284,157 @@ def apply_modifications(self):
             for separator in separators:
-                if separator in first_value and separator in second_value:
-                    return separator
+                for arg in args:
+                    if separator in arg:
+                        return separator
             return None
 
-        for x in modified_variables:
-            current = this_environment[x]
-            modified = after_source_env[x]
-            sep = return_separator_if_any(current, modified)
-            if sep is None:
-                # We just need to set the variable to the new value
-                env.set(x, after_source_env[x])
+        # Add variables to env.
+        # Assume that variables with 'PATH' in the name or that contain
+        # separators like ':' or ';' are more likely to be paths
+        for x in new_variables:
+            sep = return_separator_if_any(env_after[x])
+            if sep:
+                env.prepend_path(x, env_after[x], separator=sep)
+            elif 'PATH' in x:
+                env.prepend_path(x, env_after[x])
             else:
-                current_list = current.split(sep)
-                modified_list = modified.split(sep)
+                # We just need to set the variable to the new value
+                env.set(x, env_after[x])
+
+        for x in unset_variables:
+            env.unset(x)
+
+        for x in modified_variables:
+            before = env_before[x]
+            after = env_after[x]
+            sep = return_separator_if_any(before, after)
+            if sep:
+                before_list = before.split(sep)
+                after_list = after.split(sep)
+
+                # Filter out empty strings
+                before_list = list(filter(None, before_list))
+                after_list = list(filter(None, after_list))
+
                 # Paths that have been removed
                 remove_list = [
-                    ii for ii in current_list if ii not in modified_list]
-                # Check that nothing has been added in the middle of vurrent
-                # list
+                    ii for ii in before_list if ii not in after_list]
+                # Check that nothing has been added in the middle of
+                # before_list
                 remaining_list = [
-                    ii for ii in current_list if ii in modified_list]
-                start = modified_list.index(remaining_list[0])
-                end = modified_list.index(remaining_list[-1])
-                search = sep.join(modified_list[start:end + 1])
-
-                if search not in current:
+                    ii for ii in before_list if ii in after_list]
+                try:
+                    start = after_list.index(remaining_list[0])
+                    end = after_list.index(remaining_list[-1])
+                    search = sep.join(after_list[start:end + 1])
+                except IndexError:
+                    env.prepend_path(x, env_after[x])
```

Comment:
Maybe? You wrote the code 😝 

This isn't just a theoretical issue. Since this has been broken, I've been manually sourcing these scripts and adding the appropriate variables. I've seen a lot of:
```
PATH=/path/to/foo:/path/to/bar:
```
where the variable ends in a separator. I've also seen a lot of:
```
PATH=/path/to/foo/../../to/bar
```
but I don't _think_ this should be a problem?

I'll let you decide what is best to do. I'm personally not a fan of try/excepts. I would rather prevent the bug from popping up than handling it after it crashes. This would definitely be something worthy of unit tests though.